src/fsharp/PostInferenceChecks.fs

```diff
@@ -485,8 +509,12 @@ and CheckExprInContext (cenv:cenv) (env:env) expr (context:ByrefCallContext) =
         BindVal cenv bind.Var
         CheckExpr cenv env body
     | Expr.Const (_,m,ty) -> 
-        CheckTypePermitByrefs cenv env m ty 
-            
+        CheckTypePermitByrefs cenv env m ty
+
+    | Expr.App(Expr.Val (v,_,_),_,_,args,m) -> 
+        if cenv.reportErrors && valRefEq cenv.g v cenv.g.nameof_vref && extractNameOf args = None then
+            errorR(Error(FSComp.SR.nameofNotPermitted(), m))
```

Comment:
It's possible to output the invalid expression too?

- `"This expression does not have a name."``, it's ok
- `"Cannot extract name from expression 'f 2', blabla some hint about usage."``, it's better if not too difficult to implement